### DISCUSSION TITLE
how can i monkeypatch request from starlette

### DISCUSSION THREAD

**Discussion by rajshahdev** [encode/starlette#1688] [Q&A]
i want to mock monkeypatch the request library from starlette
and here I am using pytest

```
from starlette.requests import Request

@pytest.fixture
def user_1(monkeypatch, request: Request):
    monkeypatch.setattr(request,'scope', {"aws.event": {"requestContext":{"authorizer":{"claims":{"sub":"11bdd74a-90a1-455e-a2a8-64a0cad728ea","cognito:groups":"users"}}}}})
    # monkeypatch.setattr("app.utils.verify_sub_group.requester_group", "users")

```

but I got an error saying
```
value = '{"aws.event": {"requestContext":{"authorizer":{"claims":{"sub":"11bdd74a-90a1-455e-a2a8-64a0cad728ea","cognito:groups":"users"}}}}}'
```

or even when i pass in the test function

```
from starlette.requests import Request

def test_get_area_returns_area_of_feature(client, user_1, monkeypatch):
    monkeypatch.setattr("app.utils.verify_sub_group.requester_sub","11bdd74a-90a1-455e-a2a8-64a0cad728ea")
    monkeypatch.setattr("app.utils.verify_sub_group.requester_group","users")
```

it points to the original(actual) function

so how can I monkeypatch with request.scope from starlette

this is the function that I need to mock
```
def verify_sub_group(sub: UUID, request: Request):
    """this function returns true if the group and uuid match with cognito uuid."""

    requester_sub = request.scope['aws.event']['requestContext']['authorizer']['claims']['sub']
    requester_group = request.scope['aws.event']['requestContext']['authorizer']['claims']['cognito:groups']
```

**Comment by rajshahdev** [comment]
> i want to mock monkeypatch the request library from starlette and here I am using pytest
> 
> ```
> from starlette.requests import Request
> 
> @pytest.fixture
> def user_1(monkeypatch, request: Request):
>     monkeypatch.setattr(request,'scope', {"aws.event": {"requestContext":{"authorizer":{"claims":{"sub":"11bdd74a-90a1-455e-a2a8-64a0cad728ea","cognito:groups":"users"}}}}})
>     # monkeypatch.setattr("app.utils.verify_sub_group.requester_group", "users")
> ```
> 
> but I got an error saying
> 
> ```
> value = '{"aws.event": {"requestContext":{"authorizer":{"claims":{"sub":"11bdd74a-90a1-455e-a2a8-64a0cad728ea","cognito:groups":"users"}}}}}'
> ```
> 
> or even when i pass in the test function
> 
> ```
> from starlette.requests import Request
> 
> def test_get_area_returns_area_of_feature(client, user_1, monkeypatch):
>     monkeypatch.setattr("app.utils.verify_sub_group.requester_sub","11bdd74a-90a1-455e-a2a8-64a0cad728ea")
>     monkeypatch.setattr("app.utils.verify_sub_group.requester_group","users")
> ```
> 
> it points to the original(actual) function
> 
> so how can I monkeypatch with request.scope from starlette
> 
> this is the function that I need to mock
> 
> ```
> def verify_sub_group(sub: UUID, request: Request):
>     """this function returns true if the group and uuid match with cognito uuid."""
> 
>     requester_sub = request.scope['aws.event']['requestContext']['authorizer']['claims']['sub']
>     requester_group = request.scope['aws.event']['requestContext']['authorizer']['claims']['cognito:groups']
> ```

@Kludex I need your help here!!

**Comment by Kludex** [comment]
> this is the function that I need to mock

Then why not mock that function directly? 🤔

What you want is to call AWS on the test suite, or...?

**Reply by rajshahdev** [reply]
Hi, I made one api endpoint that fetch the sub from starlette(through JWT token).
now when I tested this endpoint it calls the actual starlette request and not mocked one.

users.py
```
from starlette.requests import Request
@router.get("/sub")
async def get_sub(request: Request) -> Dict:
        sub = request.scope["aws.event"]["requestContext"]["authorizer"]["claims"]["sub"]
```

test_users.py
```
import pytest
from starlette.requests import Request

@pytest.fixture(scope='function')
def mocked_request_scope():
    scope = {
                "type": "http",
                "http_version": "1.1",
                "method": "http_method",
                "headers": "[[k.encode(), v.encode()] for k, v in headers.items()]",
                "path": "self.resolve_asgi_path(event, context)",
                "raw_path": "self.resolve_asgi_raw_path(event, context)",
                "root_path": "self.resolve_asgi_root_path(event, context)",
                "scheme": 'headers.get("x-forwarded-proto", "https")',
                "query_string": "query_string",
                "server": "server",
                "client": "client",
                "asgi": {"version": "3.0"},
                "aws.event" : 
                            {"requestContext": 
                                {"authorizer": 
                                    {"claims": 
                                        {
                                            "sub" : "11bdd74a-90a1-455e-a2a8-64a0cad728ea"
                                        }
                                    }
                                }
                            }
                }
    return scope

@pytest.fixture
def mocked_sub(monkeypatch, mocked_request_scope):
    monkeypatch.setattr("starlette.requests.Request.scope", mocked_request_scope)

def test_get_users(mocked_sub, client):
    response = client.get("/users/sub")
    content = response.json()
    print(content)
    assert response.status_code == 200
```

right now it is not pointing to mocked_request_scope the solution will probably be specific to starlette, which as I say I don't have enough context to solve.

**Reply by rajshahdev** [reply]
i am not able to mock the function directly as I am testing it using fastapi

**Reply by rajshahdev** [reply]
sorry, the question asked earlier was not descriptive, maybe this one gives you more context on what I want to say!!

**Reply by symroe** [reply]
FWIW I have the same requirements, more or less.

I have [some middleware](https://github.com/DemocracyClub/aggregator-api/blob/e5d0c9e002824afe8e9db943f14305b3d6a267be/api/common/middleware.py#L52) that looks for AWS Lambda Authorizer data in `scope`. 

It feels like I should be able to change the content of `scope` according to the known values I'll see in Lambda. I've not found a way to do this, though. 

Exposing `scope` on `TestClient` somehow would be handy, although I'm not sure that's the right place. Either way, some interface to alter the `scope` object from inside tests would be really useful.

For my case I ended up using `Mangum` in the tests. This gives me a _fixed_ AWS-like scope, but I still can't test how my app reacts to this scope changing, so it's not an ideal situation.
